core/src/Cache/CacheFallbackOperations.cs

```diff
@@ -63,54 +63,70 @@ internal class CacheFallbackOperations
-                uniqueId, idItem.IdToken.PreferredUsername);
-            AdalResultWrapper wrapper = new AdalResultWrapper()
-            {
-                Result = new AdalResult(null, null, DateTimeOffset.MinValue)
+                //Using scope instead of resource becaue that value does not exist. STS should return it.
+                AdalTokenCacheKey key = new AdalTokenCacheKey(authority, scope, rtItem.ClientId, TokenSubjectType.User,
+                    uniqueId, idItem.IdToken.PreferredUsername);
+                AdalResultWrapper wrapper = new AdalResultWrapper()
                 {
-                    UserInfo = new AdalUserInfo()
+                    Result = new AdalResult(null, null, DateTimeOffset.MinValue)
                     {
-                        UniqueId = uniqueId,
-                        DisplayableId = idItem.IdToken.PreferredUsername
-                    }
-                },
-                RefreshToken = rtItem.Secret,
-                RawClientInfo = rtItem.RawClientInfo,
-                //ResourceInResponse is needed to treat RT as an MRRT. See IsMultipleResourceRefreshToken 
-                //property in AdalResultWrapper and its usage. Stronger design would be for the STS to return resource
-                //for which the token was issued as well on v2 endpoint.
-                ResourceInResponse = scope
-            };
+                        UserInfo = new AdalUserInfo()
+                        {
+                            UniqueId = uniqueId,
+                            DisplayableId = idItem.IdToken.PreferredUsername
+                        }
+                    },
+                    RefreshToken = rtItem.Secret,
+                    RawClientInfo = rtItem.RawClientInfo,
+                    //ResourceInResponse is needed to treat RT as an MRRT. See IsMultipleResourceRefreshToken 
+                    //property in AdalResultWrapper and its usage. Stronger design would be for the STS to return resource
+                    //for which the token was issued as well on v2 endpoint.
+                    ResourceInResponse = scope
+                };
 
-#if !FACADE && !NETSTANDARD1_3 || !NET_CORE
-            IDictionary<AdalTokenCacheKey, AdalResultWrapper> dictionary = AdalCacheOperations.Deserialize(legacyCachePersistance.LoadCache());
-            dictionary[key] = wrapper;
-            legacyCachePersistance.WriteCache(AdalCacheOperations.Serialize(dictionary));
-#endif
+                IDictionary<AdalTokenCacheKey, AdalResultWrapper> dictionary = AdalCacheOperations.Deserialize(legacyCachePersistance.LoadCache());
+                dictionary[key] = wrapper;
+                legacyCachePersistance.WriteCache(AdalCacheOperations.Serialize(dictionary));
+            }
+            catch (Exception ex)
+            {
+                string msg = "An error occurred while writing ADAL refresh token.";
```

Comment:
not sure what do you mean under ADAL-00000001, also "Continuing with-out writing the token to the cache" is not very accurate because this error is error of sharing RT in another format , so at this  point token already should be stored in "main" cache format.
I have updated messages.

Reply:
1: The ID
What I mean by the ID in front of the message it to provide a unique id for every error message we show. Thus, pls. create a data class with error id's and one with error messages. For now these can all be internal. Use those for all the new messages you are adding. Ideally you would do the same to all the messages in this file. 

2: Ensure the message indicates that it's not blocking
The message should be as precise as possible, what I was indicating is that you should add to the message that even though this is an error situation, then the situation was not throwing any exceptions and thus execution flow is not interrupted by this situation.

3: Please add and aka.ms link to the errors and ensure the link is added to the message in the log or messages sent to customers. e.g. aka.ms/msal_cache_persistance_errors.

Please stop by for additional clarifications.

---
In reply to: [210447020](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1199#discussion_r210447020) [](ancestors = 210447020)

Reply:
1. Do you mean for every possible log message ?
    These messages are not part of api they are log messages.
2. In this cases exception are not thrown, not sure why we want to say that something was not done, especially taking to account that warning message level is used 
3.  aka.ms/msal_cache_persistance_errors - not sure what to add for these cases taking to account  that we catch general exception - anything can be there

Reply:
Context: All of these log messages are happening due to some exception!

Consider that someone sends us the log we will search for the root cause in the code. ... Now consider that the user is able to lookup this information in the wiki... the developer will no longer have to open an issue or a support case to find all the information needed. The page linked to should just explain a bit about the cases where things can go wrong. If each log message/exception has an id then it will be easy for a developer to find more information about the root cause.

---
In reply to: [210739714](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1199#discussion_r210739714) [](ancestors = 210739714)